specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+
+### `WeibullThresholdConfig`
+
+Config for the `WeiBullThreshold` chunking strategy.
+
+Makes chunk boundary decisions on the hash of the key of a []byte pair and tries to create chunks that have an average number of []byte pairs, rather than an average number of Bytes. However, because the target number of []byte pairs is computed directly from the chunk size and count, the practical difference in the distribution of chunk sizes is minimal. It uses a dynamic threshold modeled on a weibull distribution (https://en.wikipedia.org/wiki/Weibull_distribution). As the size of the current trunk increases, it becomes easier to pass the threshold, reducing the likelihood of forming very large or very small chunks.
+
+The `K` value represents the Shape Parameter, and `L` represents `λ` - the scaling factor.
+
+### `RollingHashConfig`
+
+ rollingHashSplitter is a nodeSplitter that makes chunk boundary decisions using a rolling value hasher that processes Item pairs in a byte-wise fashion.
+
+rollingHashSplitter uses a dynamic hash pattern designed to constrain the chunk Size distribution by reducing the likelihood of forming very large or very small chunks. As the Size of the current chunk grows, rollingHashSplitter changes the target pattern to make it easier to match. The result is a chunk Size distribution that is closer to a binomial distribution, rather than geometric.
+
+## Algorithm in detail
+
+Left to the implementation is how to generate and load data from IPLD.
+Whenever a Link is resolved to a ProllyNode or ChunkConfig, this is done via the IPLD LinkSystem which it an implementation detail outside of the scope of this document.
+
+When serializing data into a CID+Block, one should use the codec and multihash that's either used by root CID of the tree you are modifying, or specificed explicitly during tree creation. Blocks should also be saved to the IPLD LinkSystem and how this is done is also outside the scope of this document. We will assume that there is a `getNode(cid)` API which loads IPLD Nodes from a CID, and a `saveNode(node) => {cid, bytes}` which will save a node and get back a CID and the byte contents used for generating the CID. Note that `saveNode` should use the same encoding and hash algorithm as the root of the tree.
+
+### IsLeaf(ProllyNode) : Boolean
+
+1. get the `level` of the `ProllyNode`
+2. if the `level` is `0` return `true`
+3. else return `false`
+
+### CursorIsValid(Cursor) : Boolean
+
+1. get the `length` of `cursor.node.keys`
+2. if `length` is `0`, return `false`
+3. if `cursor.index` is less than 0, return `false`
+4. if `cursor.index` is greater than or equal to `legnth`, return `false`
+
+### KeyIndex(ProllyNode, item) : Integer
+
+Use a [binary search](https://en.wikipedia.org/wiki/Binary_search_algorithm) or equivalent to find the index in the `keys` array which is "closest" to but not "larger than" the `item`. Return the index.
+
+### CursorAtItem(ProllyNode, item): Cursor
```

Comment:
nodeStore should be part of parameters for link resolving, maybe compareFunc(how to compare two []byte key) is also meaningful

Reply:
If you check out the beginning of the "Algorithm in detail" section I mention how the APIs for saving/loading nodes is intentionally left abstract so that implementations can use whatever they want.